bob/src/hw_metrics_collector.rs

```diff
@@ -179,53 +183,165 @@ enum CommandError {
+        if !self.procfs_avl {
             return Err(CommandError::Unavailable);
         }
 
-        match parse_command_output(Command::new("iostat").arg("-c")) {
-            Ok(output) => {
-                // find avg-cpu headers line
-                let mut lines = output
-                    .lines()
-                    .skip_while(|line| line.find("avg-cpu:").is_none());
-                // find iowait column on the next line
-                if let Some(line) = lines.next() {
-                    let mut values = line.split_whitespace();
-                    // iowait is in 3d column
-                    if let Some(iowait_str) = values.nth(3) {
-                        if let Ok(iowait) = Self::parse_iostat_result(iowait_str) {
-                            return Ok(iowait);
+        const CPU_IOWAIT_COLUMN: usize = 5;
+        let mut err = None;
+        match Self::stat_cpu_line() {
+            Ok(line) => {
+                let parts: Vec<&str> = line.split_whitespace().collect();
+                if parts.len() > CPU_IOWAIT_COLUMN {
+                    let mut sum = 0.;
+                    let mut f_iowait = 0.;
+                    for i in 1..parts.len() {
+                        match parts[i].parse::<f64>() {
+                            Ok(val) => {
+                                sum += val;
+                                if i == CPU_IOWAIT_COLUMN {
+                                    f_iowait = val;
+                                }
+                            },
+                            Err(_) => {
+                                let msg = format!("Can't parse {}", CPU_STAT_FILE);
+                                err = Some(msg);
+                                break;
+                            }
                         }
                     }
+                    if err.is_none() {
+                        return Ok(f_iowait * 100. / sum);
+                    }
+                } else {
+                    let msg = format!("CPU stat format in {} changed", CPU_STAT_FILE);
+                    err = Some(msg);
                 }
-            }
+            },
             Err(e) => {
-                self.iostat_avl = false;
-                return Err(CommandError::Primary(e));
+                err = Some(e);
+            }
+        }
+        self.procfs_avl = false;
+        Err(CommandError::Primary(err.unwrap()))
+    }
+}
+
+const DIFFCONTAINER_THRESHOLD: Duration = Duration::from_millis(500);
+
+struct TimeContainer {
+    last: Option<Instant>,
+    prelast: Option<Instant>,
+}
+
+enum TimeMoment {
+    Last(Duration),
+    Prelast(Duration),
+}
+
+impl TimeMoment {
+    fn elapsed_ms(&self) -> f64 {
+        match self {
+            TimeMoment::Last(dur) => dur.as_millis() as f64,
+            TimeMoment::Prelast(dur) => dur.as_millis() as f64,
+        }
+    }
+}
+
+impl TimeContainer {
+    fn new() -> Self {
+        Self {
+            last: None,
+            prelast: None,
```

Comment:
Actually, I just realized that we don't need to keep track of `prelast` values, because we don't need to provide values on every update. If we don't provide values, then metrics library will simply report a previous ones to graphite/prometheus. So the logic can be simplified: we just check, that the elapsed time since the last update is greater than `DIFFCONTAINER_THRESHOLD`. If it is then we publish new values and update the last to the current ones, if it is not then do nothing.

I apologize that I previously suggested to implement more complex logic and that now I ask you to simplify it